sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -88,5 +85,158 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+
+                return (pendingTask.EnsureCompleted(), false);
+            }
+
+            public string SaveToken(AccessToken token)
+            {
+                if (string.IsNullOrEmpty(token.Token))
+                {
+                    var exception = new InvalidOperationException($"{nameof(TokenCredential)}.{nameof(TokenCredential.GetToken)} has failed with unknown error.");
+                    ResetTokenIfExpired(exception);
+                    throw exception;
+                }
+
+                string headerValue;
+                TaskCompletionSource<string>? pendingTcs;
+
+                lock (_syncObj)
+                {
+                    _headerValue = "Bearer " + token.Token;
+                    _refreshOn = token.ExpiresOn - _tokenRefreshOffset;
+                    _expiresOn = token.ExpiresOn;
+                    _tokenState = TokenState.Valid;
+
+                     pendingTcs = _pendingTcs;
+                    headerValue = _headerValue;
+                    _pendingTcs = null;
+                }
+
+                pendingTcs?.SetResult(headerValue);
+                return headerValue;
+            }
+
+            public void ResetTokenIfExpired(Exception exception)
+            {
+                TaskCompletionSource<string>? pendingTcs = default;
+
+                lock (_syncObj)
+                {
+                    if (DateTimeOffset.UtcNow >= _expiresOn)
+                    {
+                        _headerValue = default;
+                        _refreshOn = default;
+                        _expiresOn = default;
+                        _tokenState = TokenState.Invalid;
+
+                        pendingTcs = _pendingTcs;
+                        _pendingTcs = null;
+                    }
+                    else if (_tokenState == TokenState.AboutToExpire)
+                    {
+                        _tokenState = TokenState.Valid;
+                        if (_pendingTcs != default)
+                        {
+                            exception = new InvalidOperationException($"{nameof(GetHeaderValueAsync)} shouldn't wait in {TokenState.AboutToExpire} state.", exception);
+                        }
+                    }
+                }
+
+                pendingTcs?.SetException(exception);
+            }
+
+            private (string? headerValue, Task<string>? pendingTask, bool scheduleRefreshToken) GetHeaderValue()
+            {
+                lock (_syncObj)
+                {
+                    if (DateTimeOffset.UtcNow >= _expiresOn && _tokenState != TokenState.Pending)
+                    {
+                        _tokenState = TokenState.Pending;
+                        _headerValue = default;
+                        return (default, default, false);
+                    }
+
+                    if (DateTimeOffset.UtcNow >= _refreshOn && _tokenState == TokenState.Valid)
```

Comment:
This needs to be updated to ensure we don't refresh more often than every 30 seconds, if the token is still not expired.

Reply:
Fixed, changed test